eclipse/che-theia
# Support remote file system on side car

### What does this PR do?
This change proposal adds support for remote file system to service VS Code extensions on the plugin side car.
The main concept of this change is based on the registering the custom file scheme, like `file-sidecar-${machine-name}` with the corresponding file system provider for each sidecar and intercepting any call (of [FileSystem](https://github.com/eclipse-theia/theia/blob/master/packages/plugin/src/theia.d.ts#L5380-L5449) interface) from VS Code extensions by replacing the scheme in URI objects. So that approach allows third party extensions to work properly inside cars like in one container with Theia editor.

Signed-off-by: Vladyslav Zhukovskyi <vzhukovs@redhat.com>

### Screenshot/screencast of this PR
![Eclipse Che 2021-03-22 18-56-04](https://user-images.githubusercontent.com/1968177/112027880-4d8a2900-8b40-11eb-977f-476d778e5f87.jpg)


### What issues does this PR fix or reference?
fixes https://github.com/eclipse/che/issues/19110


### How to test this PR?
Open workspace from the following devfile:
```yaml
apiVersion: 1.0.0
metadata:
  name: python-workspace
attributes:
  persistVolumes: 'false'
projects:
  - name: jupyter-hello-world
    source:
      location: 'https://github.com/chasbecker/TextAnalysis.git'
      type: git
      branch: master
components:
  - memoryLimit: 512Mi
    type: chePlugin
    reference: 'https://gist.github.com/vzhukovs/34a31b804030477cad977314509a86de/raw/9b0ae6ea71366698feea0f52dece3a343e499711/meta.yaml'
    alias: ms-python/python/latest
  - memoryLimit: 512Mi
    type: cheEditor
    reference: 'https://raw.githubusercontent.com/chepullreq4/pr-check-files/master/che-theia/pr-1045/che_theia_meta.yaml'
    alias: che-theia

```

This devfile contains the VS Code plugin(source code located [here](https://github.com/vzhukovs/remote-file-system-test), vsix file located [here](https://github.com/vzhukovs/remote-file-system-test/raw/main/remote-file-system-test-0.0.1.vsix)).

There are two ways how to check this PR:

1. As far as webviews from vscode-python extension use VS Code file system, it is enough to open any `.ipynb` file and to ensure, that python playbook has visual representation as in the above screenshot.
2. Call command `Remote File System - Show Case` from test extension. This extension uses the VS Code file system under hood and this extension is located in dedicate side car as far as vscode-python extension.
    - Extension calls the following VS Code methods which described in [extension.ts#L17-L51](https://github.com/vzhukovs/remote-file-system-test/blob/main/src/extension.ts#L17-L51)


### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### Happy Path Channel

HAPPY_PATH_CHANNEL=stable
